From 7b5ee97196637922a4f565ab5f28c15c25f26ea3 Mon Sep 17 00:00:00 2001
From: Guilhem Bichot <guilhem.bichot@contributor>


Date: Wed, 18 Mar 2026 17:55:54 +0100
Subject: [PATCH] BUGFIX crash in
 JOIN::compare_costs_of_subquery_strategies

While working on this bug, I discovered another problem, also related
to aggregates vs left join elimination.

Item_subselect::compile() does not dive into its underlying subquery.
So, we cannot reliably find, in a subquery Item, what the aggregates
really depend on.

Fix: skip elimination in this case. Note that this is not about the
plain and common:
select aggregate(...) from t1 left join t2 ...
neither about
select (select aggregate-in-subq(...) ...) from t1 left join t2 ...
but rather about the less common:
select (select aggregate-in-top-q(...) ...) from t1 left join t2 ...
where 'aggregate' has the left-join query as aggregation query.
---
 mysql-test/include/join_elimination.inc       |  7 ++++++
 mysql-test/r/join_elimination.result          | 25 +++++++++++++++++++
 .../r/join_elimination-dstore.result          | 25 +++++++++++++++++++
 sql/sql_join_elimination.cc                   | 16 +++++++++++-
 4 files changed, 72 insertions(+), 1 deletion(-)

diff --git a/mysql-test/include/join_elimination.inc b/mysql-test/include/join_elimination.inc
index c352acb32ab..b0c96c5fba4 100644
--- a/mysql-test/include/join_elimination.inc
+++ b/mysql-test/include/join_elimination.inc
@@ -703,6 +703,13 @@ SELECT COUNT(1) FROM
 eval EXPLAIN $query;
 eval $query;
 
+--echo # Fix for a bug: it's not possible to get reliable used_tables information of an aggregate which is inside a subquery of our query block. We block this situation in code, otherwise we would not notice that t2 is used in count() and eliminate it, which would be wrong.
+explain select distinct (select count(t2.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+--echo # It's also not possible to set reliable used_tables information after eliminating; so we cannot eliminate t1 here; count() would still pretend it depends on t2.
+explain select distinct (select count(t1.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+--echo # Here the aggregate in the subquery, is not an aggregate from the POV of the query block where the left join is, so there is no problem.
+explain select distinct (select count(t0.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+
 --echo # Without GROUP BY, and with composite unique key:
 
 CREATE TABLE u (a INT, b INT, c INT NOT NULL, UNIQUE (a,b));
diff --git a/mysql-test/r/join_elimination.result b/mysql-test/r/join_elimination.result
index aba30539612..5ffb10f0d19 100644
--- a/mysql-test/r/join_elimination.result
+++ b/mysql-test/r/join_elimination.result
@@ -1815,6 +1815,31 @@ GROUP BY t.a, computed_col
 ) AS derived;
 COUNT(1)
 1
+# Fix for a bug: it's not possible to get reliable used_tables information of an aggregate which is inside a subquery of our query block. We block this situation in code, otherwise we would not notice that t2 is used in count() and eliminate it, which would be wrong.
+explain select distinct (select count(t2.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
+1	PRIMARY	t1	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+1	PRIMARY	t2	NULL	eq_ref	PRIMARY	PRIMARY	4	test.t1.a	1	100.00	Using index
+2	DEPENDENT SUBQUERY	t0	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+Warnings:
+Note	1276	Field or reference 'test.t2.a' of SELECT #2 was resolved in SELECT #1
+Note	1003	/* select#1 */ select (/* select#2 */ select count((`test`.`t2`.`a` + 0)) from `test`.`t1` `t0`) AS `(select count(t2.a+0) from t1 t0)` from `test`.`t1` left join `test`.`t1` `t2` on((`test`.`t2`.`a` = `test`.`t1`.`a`)) where true
+# It's also not possible to set reliable used_tables information after eliminating; so we cannot eliminate t1 here; count() would still pretend it depends on t2.
+explain select distinct (select count(t1.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
+1	PRIMARY	t1	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+1	PRIMARY	t2	NULL	eq_ref	PRIMARY	PRIMARY	4	test.t1.a	1	100.00	Using index
+2	DEPENDENT SUBQUERY	t0	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+Warnings:
+Note	1276	Field or reference 'test.t1.a' of SELECT #2 was resolved in SELECT #1
+Note	1003	/* select#1 */ select (/* select#2 */ select count((`test`.`t1`.`a` + 0)) from `test`.`t1` `t0`) AS `(select count(t1.a+0) from t1 t0)` from `test`.`t1` left join `test`.`t1` `t2` on((`test`.`t2`.`a` = `test`.`t1`.`a`)) where true
+# Here the aggregate in the subquery, is not an aggregate from the POV of the query block where the left join is, so there is no problem.
+explain select distinct (select count(t0.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
+1	PRIMARY	t1	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+2	SUBQUERY	t0	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+Warnings:
+Note	1003	/* select#1 */ select distinct (/* select#2 */ select count((`test`.`t0`.`a` + 0)) from `test`.`t1` `t0`) AS `(select count(t0.a+0) from t1 t0)` from `test`.`t1`
 # Without GROUP BY, and with composite unique key:
 CREATE TABLE u (a INT, b INT, c INT NOT NULL, UNIQUE (a,b));
 # Only one key part is functionally dependent, no elimination:
diff --git a/mysql-test/suite/dstore_main/r/join_elimination-dstore.result b/mysql-test/suite/dstore_main/r/join_elimination-dstore.result
index c6aab03e095..8a271885d19 100644
--- a/mysql-test/suite/dstore_main/r/join_elimination-dstore.result
+++ b/mysql-test/suite/dstore_main/r/join_elimination-dstore.result
@@ -1815,6 +1815,31 @@ GROUP BY t.a, computed_col
 ) AS derived;
 COUNT(1)
 1
+# Fix for a bug: it's not possible to get reliable used_tables information of an aggregate which is inside a subquery of our query block. We block this situation in code, otherwise we would not notice that t2 is used in count() and eliminate it, which would be wrong.
+explain select distinct (select count(t2.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
+1	PRIMARY	t1	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+1	PRIMARY	t2	NULL	eq_ref	PRIMARY	PRIMARY	4	test.t1.a	1	100.00	Using index
+2	DEPENDENT SUBQUERY	t0	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+Warnings:
+Note	1276	Field or reference 'test.t2.a' of SELECT #2 was resolved in SELECT #1
+Note	1003	/* select#1 */ select (/* select#2 */ select count((`test`.`t2`.`a` + 0)) from `test`.`t1` `t0`) AS `(select count(t2.a+0) from t1 t0)` from `test`.`t1` left join `test`.`t1` `t2` on((`test`.`t2`.`a` = `test`.`t1`.`a`)) where true
+# It's also not possible to set reliable used_tables information after eliminating; so we cannot eliminate t1 here; count() would still pretend it depends on t2.
+explain select distinct (select count(t1.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
+1	PRIMARY	t1	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+1	PRIMARY	t2	NULL	eq_ref	PRIMARY	PRIMARY	4	test.t1.a	1	100.00	Using index
+2	DEPENDENT SUBQUERY	t0	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+Warnings:
+Note	1276	Field or reference 'test.t1.a' of SELECT #2 was resolved in SELECT #1
+Note	1003	/* select#1 */ select (/* select#2 */ select count((`test`.`t1`.`a` + 0)) from `test`.`t1` `t0`) AS `(select count(t1.a+0) from t1 t0)` from `test`.`t1` left join `test`.`t1` `t2` on((`test`.`t2`.`a` = `test`.`t1`.`a`)) where true
+# Here the aggregate in the subquery, is not an aggregate from the POV of the query block where the left join is, so there is no problem.
+explain select distinct (select count(t0.a+0) from t1 t0) from t1 left join t1 t2 on t1.a=t2.a;
+id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
+1	PRIMARY	t1	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using index
+2	SUBQUERY	t0	NULL	index	NULL	PRIMARY	4	NULL	3	100.00	Using aggregate; Using index
+Warnings:
+Note	1003	/* select#1 */ select distinct (/* select#2 */ select count((`test`.`t0`.`a` + 0)) from `test`.`t1` `t0`) AS `(select count(t0.a+0) from t1 t0)` from `test`.`t1`
 # Without GROUP BY, and with composite unique key:
 CREATE TABLE u (a INT, b INT, c INT NOT NULL, UNIQUE (a,b));
 # Only one key part is functionally dependent, no elimination:
diff --git a/sql/sql_join_elimination.cc b/sql/sql_join_elimination.cc
index 5601244d95a..e7124122d92 100644
--- a/sql/sql_join_elimination.cc
+++ b/sql/sql_join_elimination.cc
@@ -770,7 +770,21 @@ void Query_block::do_join_elimination_for_list(
                 the query above. We add no bits for '+', go down to '2' and
                 add bits for it (none actually as it's a literal), go to
                 COUNT and add no bits, go down to '*' and add bits for it
-                and do not go down to table1.col. */
+                and do not go down to table1.col.
+
+                We have a special case: if the window function or aggregate is
+                itself in a subquery of the current Query_block, we will not
+                be able to reach to its arguments, because
+                Item_subselect::compile() does not dive into its underlying
+                Query_block's items; moreover,
+                Item_subselect::update_used_tables(), which we would need to
+                call further down after actual elimination, doesn't dive
+                either. We declare that elimination is not possible in this
+                case, by pretending that all tables are referenced. */
+                if (inner_item->type() == Item::SUBSELECT_ITEM) {
+                  ut_map = ~table_map{0};
+                  return false;
+                }
                 return true;  // Reach to arguments.
               } else {
                 // used_tables() is reliable => fast path.
-- 
2.23.0

